### PR TITLE
Doc: AssemblyBuilder require fully trusted environment with trusted input

### LINKED ISSUE #9998: Add security-related documentation to AssemblyBuilder and PersistedAssemblyBuilder
Per offline discussion, we need to add text to make it clear that AssemblyBuilder and PersistedAssemblyBuilder require a full trust environment with trusted input, similarly as other technologies such as compilers. This is necessary for several somewhat-obvious reasons:
- IL and metadata can contain secrets provided by the consumer.
- IL can be invalid such as not having a balanced push vs. pop opcodes, or invalid operands for a given opcode.
- Can contain any code that may, for example, deadlock, have infinite stack recursion, or have an infinite loop.
- IL can load and execute code on any other reachable assembly.

### PR DESCRIPTION
**Pull request by buyaa-n**
## Summary

AssemblyBuilder and PersistedAssemblyBuilder require a full trust environment with trusted input, similarly as other technologies such as compilers.

Fixes #9998

### CODE REVIEW

File: xml/System.Reflection.Emit/AssemblyBuilder.xml [hunk 1]
```diff
@@ -76,7 +76,18 @@
...
     <summary>Defines and represents a dynamic assembly.</summary>
-    <remarks>For more information about this API, see <see href="/dotnet/fundamentals/runtime-libraries/system-reflection-emit-assemblybuilder">Supplemental API remarks for AssemblyBuilder</see>.</remarks>
+    <remarks>
+      <format type="text/markdown"><![CDATA[
+
+## Remarks
+
+For more information about this API, see [Supplemental API remarks for AssemblyBuilder](/dotnet/fundamentals/runtime-libraries/system-reflection-emit-assemblybuilder).
+
+> [!WARNING]
```
Comment (buyaa-n) [inline xml/System.Reflection.Emit/AssemblyBuilder.xml:86]: Wonder if it should be `Warning` or `Note` or ...

Comment (carlossanlop) [reply xml/System.Reflection.Emit/AssemblyBuilder.xml:86]: I think Warning is good enough as it shows up in the preview: https://review.learn.microsoft.com/en-us/dotnet/api/system.reflection.emit.persistedassemblybuilder?view=net-9.0&branch=pr-en-us-10541

File: xml/System.Reflection.Emit/AssemblyBuilder.xml [hunk 2]
```diff
@@ -76,7 +76,18 @@
...
+> `AssemblyBuilder` APIs require a fully trusted environment with trusted input, similarly as other technologies such as compilers. There are no restrictions other than basic validation for generated IL, for member name, count and associated metadata such as custom attributes, that the AssemblyBuilder can contain.
```
Comment (gewarren) [inline xml/System.Reflection.Emit/AssemblyBuilder.xml:87]:
```suggestion
> `AssemblyBuilder` APIs require a fully trusted environment with trusted input, similar to other technologies such as compilers. There are no restrictions other than basic validation for generated IL that the `AssemblyBuilder` can contain. That includes validation for member name, count, and associated metadata, such as custom attributes.
```

Comment (gewarren) [reply xml/System.Reflection.Emit/AssemblyBuilder.xml:87]: Does this convey the same message?

File: xml/System.Reflection.Emit/ILGenerator.xml [hunk 3]
```diff
@@ -84,6 +84,14 @@
...
  MSIL is used as input to a just-in-time (JIT) compiler.
 
+> [!WARNING]
+> There are no restrictions, other than basic validation for ILGenerator APIs that used for producing method IL. If untrusted input is used for producing IL:
+> - IL and metadata can contain secrets provided by the consumer.
+> - IL can be invalid such as not having a balanced push vs. pop opcodes, or invalid operands for a given opcode.
+> - Can contain any code that may, for example, deadlock, have infinite stack recursion, or have an infinite loop.
+> - IL can load and execute code on any other reachable assembly.
+>
+> Such code can do whatever it wishes when the method is loaded and executed. To restrict such vulnerabilities `ILGenerator` require a fully trusted environment with trusted input.
```
Comment (gewarren) [inline xml/System.Reflection.Emit/ILGenerator.xml:94]:
```suggestion
> There are no restrictions, other than basic validation, for ILGenerator APIs that produce method IL. If untrusted input is used for producing IL:
>
> - IL and metadata can contain secrets provided by the consumer.
> - IL can be invalid, such as not having balanced push versus pop opcodes, or invalid operands for a given opcode.
> - IL can contain any code that might, for example, deadlock, have infinite stack recursion, or have an infinite loop.
> - IL can load and execute code on any other reachable assembly.
>
> Such code can do whatever it wishes when the method is loaded and executed. To protect against such vulnerabilities, `ILGenerator` requires a fully trusted environment with trusted input.
```

File: xml/System.Reflection.Emit/PersistedAssemblyBuilder.xml [hunk 4]
```diff
@@ -25,7 +25,17 @@
...
-    <remarks>To be added.</remarks>
+    <remarks>
+      <format type="text/markdown"><![CDATA[
+
+## Remarks
+
+For more information about this API, see [Persisted dynamic assemblies in .NET](/dotnet/fundamentals/runtime-libraries/system-reflection-emit-persistedassemblybuilder).
+
+> [!WARNING]
+> `PersistedAssemblyBuilder` APIs require a fully trusted environment with trusted input, similarly as other technologies such as compilers. There are no restrictions other than basic validation for generated IL, for member name, count and associated metadata such as custom attributes, that the AssemblyBuilder can contain.
```
Comment (gewarren) [inline xml/System.Reflection.Emit/PersistedAssemblyBuilder.xml:36]: Same updates as first comment.

File: xml/System.Reflection.Emit/AssemblyBuilder.xml (hunk 1 above)
Comment (buyaa-n) [reply xml/System.Reflection.Emit/AssemblyBuilder.xml:86]: Yeah saw that, still not sure Warning   
image
or Note
image

File: xml/System.Reflection.Emit/AssemblyBuilder.xml (hunk 2 above)
Comment (buyaa-n) [reply xml/System.Reflection.Emit/AssemblyBuilder.xml:87]: Yes, thanks!

File: xml/System.Reflection.Emit/PersistedAssemblyBuilder.xml (hunk 4 above)
Comment (buyaa-n) [reply xml/System.Reflection.Emit/PersistedAssemblyBuilder.xml:36]:
```suggestion
> `PersistedAssemblyBuilder` APIs require a fully trusted environment with trusted input, similar to other technologies such as compilers. There are no restrictions other than basic validation for generated IL that the `AssemblyBuilder` can contain. That includes validation for member name, count, and associated metadata, such as custom attributes.
```

File: xml/System.Reflection.Emit/AssemblyBuilder.xml (hunk 1 above)
Comment (gewarren) [reply xml/System.Reflection.Emit/AssemblyBuilder.xml:86]: I think warning is appropriate.

Comment (buyaa-n) [reply xml/System.Reflection.Emit/AssemblyBuilder.xml:86]: OK, warning then

**Review by gewarren — APPROVED** [review]

**Pull request merged**